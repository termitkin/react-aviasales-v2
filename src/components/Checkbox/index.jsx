import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { device } from "../mediaQueries";
import checkBoxIcon from "./check.svg";

const StyledLabel = styled.label`
  width: 100%;
  display: flex;
  align-items: center;
  cursor: pointer;
  font-family: Open Sans, sans-serif;
  font-weight: normal;
  font-size: 13px;
  line-height: 35px;
  color: #4a4a4a;

  &::before {
    content: "";
    display: inline-block;
    width: 19px;
    height: 19px;
    border: 1px solid #d2d5d6;
    border-radius: 3px;
    box-sizing: border-box;
    margin-right: 10px;
  }
`;

const StyledCheckbox = styled.input`
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  border: 0;
  padding: 0;
  clip: rect(0 0 0 0);
  overflow: hidden;

  &:checked + ${StyledLabel}::before {
    border: 1px solid #2196f3;
    background-image: url(${checkBoxIcon});
    background-position: center center;
    background-repeat: no-repeat;
  }

  &:focus + ${StyledLabel}::before {
    border: 3px solid #2196f3;
  }
`;

const StyledOnlyButton = styled.button`
  width: 49px;
  background-color: transparent;
  border: none;
  margin-left: auto;
  font-family: Open Sans, sans-serif;
  font-weight: 600;
  font-size: 11px;
  line-height: 35px;
  letter-spacing: 0.5px;
  color: #2196f3;
  text-transform: uppercase;
  display: none;
  cursor: pointer;
  padding: 0;
`;

const StyledCheckboxWrap = styled.div`
  width: 290px;
  height: 36px;
  padding: 0 15px;
  display: flex;
  align-items: center;
  box-sizing: border-box;
  outline: none;

  @media ${device.desktop} {
    width: 232px;
  }

  &:hover {
    background-color: #f1fcff;
  }

  &:hover ${StyledOnlyButton} {
    display: table-cell;
    background-color: #f1fcff;
  }
`;

const Checkbox = props => {
  const handleClickOnCheckbox = () => {
    return props.changeStops({
      clickedButton: "checkbox" + props.id,
      only: false
    });
  };

  const handleClickOnOnlyButton = () => {
    return props.changeStops({
      clickedButton: "checkbox" + props.id,
      only: true
    });
  };

  const onlyButton =
    props.stops === "all" ? (
      ""
    ) : (
      <StyledOnlyButton onClick={handleClickOnOnlyButton}>
        только
      </StyledOnlyButton>
    );

  return (
    <StyledCheckboxWrap tabIndex="-1">
      <StyledCheckbox
        checked={props.isEnabled}
        id={"checkbox" + props.id}
        type="checkbox"
        onChange={handleClickOnCheckbox}
      />
      <StyledLabel htmlFor={"checkbox" + props.id}>
        {props.labelText}
      </StyledLabel>
      {onlyButton}
    </StyledCheckboxWrap>
  );
};

export default Checkbox;

Checkbox.propTypes = {
  changeStops: PropTypes.func.isRequired,
  stops: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  id: PropTypes.number.isRequired,
  isEnabled: PropTypes.bool.isRequired,
  labelText: PropTypes.string.isRequired
};
