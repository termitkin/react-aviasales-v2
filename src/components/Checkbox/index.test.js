import React from "react";
import { create, act } from "react-test-renderer";
import Checkbox from "./index";
import { StyledOnlyButton } from "./index";
import { matchers } from "jest-emotion";
expect.extend(matchers);

describe("Checkbox component", () => {
  let wrapper;

  afterAll(() => {
    wrapper.unmount();
  });

  act(() => {
    wrapper = create(
      <Checkbox
        id={1}
        changeStops={() => {}}
        labelText="1 пересадка"
        isEnabled={true}
      />
    );
  });

  let tree = wrapper.toJSON();

  it("Matches the snapshot", () => {
    expect(tree).toMatchSnapshot();
  });

  it("Change background-color on hover", () => {
    expect(tree).toHaveStyleRule("background-color", "#f1fcff", {
      target: ":hover"
    });
  });

  it("Correct props values", () => {
    const instance = wrapper.root;
    const checkbox = instance.findByType("input");
    expect(checkbox.props.id).toEqual("checkbox1");
    expect(checkbox.props.type).toEqual("checkbox");
    expect(checkbox.props.checked).toEqual(true);
  });

  it("Correct label text", () => {
    const instance = wrapper.root;
    const label = instance.findByType("label");
    expect(label.children).toEqual(["1 пересадка"]);
  });

  it("Correct display value of 'only button'", () => {
    const target = { target: `:hover ${StyledOnlyButton}` };
    expect(tree).toHaveStyleRule("display", "table-cell", {
      ...target
    });
    expect(tree).toHaveStyleRule("background-color", "#f1fcff", {
      ...target
    });
  });
});
