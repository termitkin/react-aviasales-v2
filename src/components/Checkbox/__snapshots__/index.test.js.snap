// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component Matches the snapshot 1`] = `
.emotion-6 {
  width: 290px;
  height: 36px;
  padding: 0 15px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  outline: none;
}

@media (min-width:960px) {
  .emotion-6 {
    width: 232px;
  }
}

.emotion-6:hover {
  background-color: #f1fcff;
}

.emotion-6:hover .emotion-5 {
  display: table-cell;
  background-color: #f1fcff;
}

.emotion-0 {
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  border: 0;
  padding: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  overflow: hidden;
}

.emotion-0:checked + .emotion-3::before {
  border: 1px solid #2196f3;
  background-image: url(check.svg);
  background-position: center center;
  background-repeat: no-repeat;
}

.emotion-0:focus + .emotion-3::before {
  border: 3px solid #2196f3;
}

.emotion-2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  font-family: Open Sans,sans-serif;
  font-weight: normal;
  font-size: 13px;
  line-height: 35px;
  color: #4a4a4a;
}

.emotion-2::before {
  content: "";
  display: inline-block;
  width: 19px;
  height: 19px;
  border: 1px solid #d2d5d6;
  border-radius: 3px;
  box-sizing: border-box;
  margin-right: 10px;
}

.emotion-4 {
  width: 49px;
  background-color: transparent;
  border: none;
  margin-left: auto;
  font-family: Open Sans,sans-serif;
  font-weight: 600;
  font-size: 11px;
  line-height: 35px;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  color: #2196f3;
  text-transform: uppercase;
  display: none;
  cursor: pointer;
  padding: 0;
}

.emotion-4:hover,
.emotion-4:focus {
  color: #ff9d1b;
}

<div
  className="emotion-6 emotion-7"
  tabIndex="-1"
>
  <input
    checked={true}
    className="emotion-0 emotion-1"
    id="checkbox1"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="emotion-2 emotion-3"
    htmlFor="checkbox1"
  >
    1 пересадка
  </label>
  <button
    className="emotion-4 emotion-5"
    onClick={[Function]}
  >
    только
  </button>
</div>
`;
